actions/src/openstack_check_actions.py

```diff
@@ -0,0 +1,447 @@
+                    max_port=max_port,
+                    min_port=min_port,
+                    ip_prefix=ip_prefix,
+                )
+                rules_with_issues["server_list"].extend(output)
+        else:
+            output = self.check_project(
+                project=project_id,
+                cloud=cloud_account,
+                max_port=max_port,
+                min_port=min_port,
+                ip_prefix=ip_prefix,
+            )
+            rules_with_issues["server_list"] = output
+
+        return rules_with_issues
+
+    def deleting_machines_check(
+        self, cloud_account: str, project_id=None, all_projects=False
+    ):
+        """
+        Action to check for machines that are stuck deleting for more than 10mins
+        Outputs a suitable dictionary to pass into create_tickets
+        """
+        output = {
+            "title": "Server {p[id]} has not been updated in more than 10 minutes during {p[action]}",
+            "body": "The following information may be useful\nHost id: {p[id]} \n{p[data]}",
+            "server_list": [],
+        }
+
+        if all_projects:
+            with OpenstackConnection(cloud_name=cloud_account) as conn:
+                projects = conn.list_projects()
+            for project in projects:
+                deleted = self.check_deleted(project=project["id"], cloud=cloud_account)
+            output["server_list"].extend(deleted)
+        else:
+            deleted = self.check_deleted(project=project_id, cloud=cloud_account)
+            output["server_list"] = deleted
+
+        return output
+
+    def check_deleted(self, cloud: str, project: str):
+        """
+        Runs the check for deleting machines on a per-project basis
+        """
+        output = []
+
+        print("Checking project", project)
+        with OpenstackConnection(cloud_name=cloud) as conn:
+            server_list = conn.list_servers(
+                filters={"all_tenants": True, "project_id": project}
+            )
+        # Loop through each server in the project/list
+        for i in server_list:
+            # Take current time and check difference between updated time
+            since_update = datetime.utcnow() - datetime.strptime(
+                i["updated"], "%Y-%m-%dT%H:%M:%Sz"
+            )
+            # Check if server has been stuck in deleting for too long.
+            # (uses the last updated time so if changes have been made
+            # to the server while deleting the check may not work.)
+            if i["status"] == "deleting" and since_update.total_seconds() >= 600:
+                # Append data to output array
+                output.append(
+                    {
+                        "dataTitle": {"id": str(i.id), "action": str(i["status"])},
+                        "dataBody": {"id": i["id"], "data": json.dumps(i)},
+                    }
+                )
+
+        return output
+
+    def loadbalancer_check(self, cloud_account: str):
```

Comment:
Ditto on this method doing too much in a single body, can we break it down into sub methods?

Reply:
I've taken out a little bit of this. This method ends up looking pretty unwieldy but most of it is just generating the different metadata for tickets for each variation, so I don't think I can split up any more than I have in the commit I just did.